# Домашнее задание к лекции "Базовые понятия статистики"

## Обязательная часть

Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 

### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

### Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [2]:
import pandas as pd
import numpy as np

### Задание 1. Базовое изучение
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [3]:
horse = pd.read_csv('C:\\Users\\2-hom\\Documents\\Data_analysis\\Pyton_Notebook\\Статистика\\практика_и_дз\\horse_data.csv', na_values=['?'])
horse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 28 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   2        298 non-null    float64
 1   1        299 non-null    int64  
 2   530101   299 non-null    int64  
 3   38.50    239 non-null    float64
 4   66       275 non-null    float64
 5   28       241 non-null    float64
 6   3        243 non-null    float64
 7   3.1      230 non-null    float64
 8   ?        253 non-null    float64
 9   2.1      267 non-null    float64
 10  5        244 non-null    float64
 11  4        255 non-null    float64
 12  4.1      243 non-null    float64
 13  ?.1      196 non-null    float64
 14  ?.2      194 non-null    float64
 15  ?.3      53 non-null     float64
 16  3.2      197 non-null    float64
 17  5.1      181 non-null    float64
 18  45.00    270 non-null    float64
 19  8.40     266 non-null    float64
 20  ?.4      135 non-null    float64
 21  ?.5      102 non

In [4]:
horse.head()

2  1   530101  38.50     66    28    3  3.1    ?  2.1  ...  45.00  8.40  \
0  1.0  1   534817   39.2   88.0  20.0  NaN  NaN  4.0  1.0  ...   50.0  85.0   
1  2.0  1   530334   38.3   40.0  24.0  1.0  1.0  3.0  1.0  ...   33.0   6.7   
2  1.0  9  5290409   39.1  164.0  84.0  4.0  1.0  6.0  2.0  ...   48.0   7.2   
3  2.0  1   530255   37.3  104.0  35.0  NaN  NaN  6.0  2.0  ...   74.0   7.4   
4  2.0  1   528355    NaN    NaN   NaN  2.0  1.0  3.0  1.0  ...    NaN   NaN   

   ?.4  ?.5  2.2  2.3  11300  00000  00000.1  2.4  
0  2.0  2.0  3.0    2   2208      0        0    2  
1  NaN  NaN  1.0    2      0      0        0    1  
2  3.0  5.3  2.0    1   2208      0        0    1  
3  NaN  NaN  2.0    2   4300      0        0    2  
4  NaN  NaN  1.0    2      0      0        0    2  

[5 rows x 28 columns]

In [5]:
horse.rename(columns={'2': '1.Операция', '1': '2.Возраст', '530101':'3.Больницa', '38.50':'4.T_рек','66':'5.Пульс', '28':'6.Дыхание',
                     '3':'7.T_кон', '3.1':'8.Пульс_п', '?':'9.Слизистая', '2.1':'10.t _капилляры', '5':'11.Боль', '4':'12.Перистальтика',
                     '4.1':'13.Вздутие', '?.1':'14.Зонд', '?.2':'15.Рефлюкс','?.3':'16.Рефлюкс_PH', '3.2':'17.Кал', '5.1':'18.Живот',
                     '45.00':'19.Эритроциты', '8.40':'20.Об.белок_1', '?.4':'21.Бр.полость', '?.5':'22.Об.белок_2', '2.2':'23.Результат',
                     '2.3':'24.Х_поражение', '11300':'25.Тип_пораж1', '00000':'26.Тип_пораж2', '00000.1':'27.Тип_пораж3', '2.4':'28.Паталогии'},
                     inplace=True)
horse.head()


1.Операция  2.Возраст  3.Больницa  4.T_рек  5.Пульс  6.Дыхание  7.T_кон  \
0         1.0          1      534817     39.2     88.0       20.0      NaN   
1         2.0          1      530334     38.3     40.0       24.0      1.0   
2         1.0          9     5290409     39.1    164.0       84.0      4.0   
3         2.0          1      530255     37.3    104.0       35.0      NaN   
4         2.0          1      528355      NaN      NaN        NaN      2.0   

   8.Пульс_п  9.Слизистая  10.t _капилляры  ...  19.Эритроциты  20.Об.белок_1  \
0        NaN          4.0              1.0  ...           50.0           85.0   
1        1.0          3.0              1.0  ...           33.0            6.7   
2        1.0          6.0              2.0  ...           48.0            7.2   
3        NaN          6.0              2.0  ...           74.0            7.4   
4        1.0          3.0              1.0  ...            NaN            NaN   

   21.Бр.полость  22.Об.белок_2  23.Результат  24.Х_поражение  25.Тип_пораж1  \
0            2.0            2.0           3.0               2           2208   
1            NaN            NaN           1.0               2              0   
2            3.0            5.3           2.0               1           2208   
3            NaN            NaN           2.0               2           4300   
4            NaN            NaN           1.0               2              0   

   26.Тип_пораж2  27.Тип_пораж3  28.Паталогии  
0              0              0             2  
1              0              0             1  
2              0              0             1  
3              0              0             2  
4              0              0             2  

[5 rows x 28 columns]

Для анализа выбраны следующие показатели:
- 1: операция?
          1 = Да, была операция
          2 = Лечение прошло без хирургического вмешательства

- 2: Возраст
          1 = взрослая лошадь
          2 = Молодой (<6 месяцев)
          
- 4: ректальная температура
          - линейный
          - в градусах Цельсия.
          - Повышенная температура может возникнуть из-за инфекции.
          - возможно снижение температуры при позднем шоке.
          - нормальная температура 37,8
          - этот параметр обычно меняется по мере развития проблемы
               например. может начаться нормально, а затем повыситься из-за
                   поражение, возвращаясь через нормальный диапазон, как
                   лошадь в шоке 
- 5: пульс
          - линейный
          - частота пульса в ударах в минуту
          - отражает состояние сердца: 30-40 - нормально для взрослых
          - редко бывает ниже нормы, хотя у спортивных лошадей
            может иметь коэффициент 20-25
          - животные с болезненными поражениями или страдающие от шока кровообращения
            может иметь повышенную частоту сердечных сокращений
            
- 9: слизистые оболочки
          - субъективное измерение цвета
          - возможные значения:
               1 = нормальный розовый
               2 = ярко-розовый
               3 = бледно-розовый
               4 = бледно-синюшный
               5 = ярко-красный / впрыснутый
               6 = темно-синюшный
          - 1 и 2, вероятно, указывают на нормальное или слегка повышенное
            обращение
          - 3 могут возникнуть при раннем шоке
          - 4 и 6 указывают на серьезное нарушение кровообращения
          - 5 больше указывает на сепсис
          
- 13: вздутие живота
          - ВАЖНЫЙ параметр.
          - возможные значения
               1 = нет
               2 = легкий
               3 = умеренный
               4 = тяжелая
          - животное с вздутием живота может причинить боль и
            имеют сниженную перистальтику кишечника.
          - лошади с сильным вздутием живота, вероятно, потребуются
            операция просто сними давление
            
- 16: назогастральный рефлюкс, PH
          - линейный
          - шкала от 0 до 14, где 7 нейтрально.
          - нормальные значения находятся в диапазоне от 3 до 4
          
- 22: общий белок брюшной полости
          - линейный
          - чем выше уровень протеина, тем больше вероятность
            скомпрометированный кишечник. Значения указаны в гмс / дл.

- 23: результат
          - что в итоге случилось с лошадью?
          - возможные значения:
               1 = жил
               2 = умер
               3 = был усыплен

In [23]:
## расчет базовых метрик:

## min, max, размах по количественным данным (столбцы 4,5,16,22 )

print('Ректальная температура, max: ', horse['4.T_рек'].max())
print('Ректальная температура, min: ', horse['4.T_рек'].min())
print('Размах ректальной температуры: ', horse['4.T_рек'].max() - horse['4.T_рек'].min())

print('Пульс, max: ', horse['5.Пульс'].max())
print('Пульс, min: ', horse['5.Пульс'].min())
print('Размах пульса: ', horse['5.Пульс'].max() - horse['5.Пульс'].min())


print('Назогастральный рефлюкс, PH, max: ', horse['16.Рефлюкс_PH'].max())
print('Назогастральный рефлюкс, PH, min: ', horse['16.Рефлюкс_PH'].min())
print('Размах Назогастральный рефлюкс, PH: ', horse['16.Рефлюкс_PH'].max() - horse['16.Рефлюкс_PH'].min())

print('Общий белок брюшной полости, max: ', horse['22.Об.белок_2'].max())
print('Общий белок брюшной полости, min: ', horse['22.Об.белок_2'].min())
print('Размах Общий белок брюшной полости, max: ', horse['22.Об.белок_2'].max() - horse['22.Об.белок_2'].min())

Ректальная температура, max:  40.8
Ректальная температура, min:  35.4
Размах ректальной температуры:  5.399999999999999
Пульс, max:  184.0
Пульс, min:  30.0
Размах пульса:  154.0
Назогастральный рефлюкс, PH, max:  7.5
Назогастральный рефлюкс, PH, min:  1.0
Размах Назогастральный рефлюкс, PH:  6.5
Общий белок брюшной полости, max:  10.1
Общий белок брюшной полости, min:  0.1
Размах Общий белок брюшной полости, max:  10.0


Значения ректальной температуры при кишечных коликах у лошадей колеблется от критически низких температур при позднем шоке (min 35.4) до критически высоких (max 40.8), при достижении 41,7°C состояние животного переходит в критическое. Если мгновенно не отреагировать, то лошадь умрет.

Аналогичный большой размах значений у пульса: от нормального значения (30.0) до критически высокого - 184

Показатели назогастрального рефлюкса (PH) являются нейтральными (шкала от 0 до 14, где 7 нейтрально, нормальные значения находятся в диапазоне от 3 до 4).

Общий белок брюшной полости также показывает большой размах.

In [24]:
## среднее арифметическое

print('Cреднее арифметическое, Ректальная температура: ', horse['4.T_рек'].mean())
print('Cреднее арифметическое, Пульс: ', horse['5.Пульс'].mean())
print('Cреднее арифметическое, Назогастральный рефлюкс, PH: ', horse['16.Рефлюкс_PH'].mean())
print('Cреднее арифметическое, Общий белок брюшной полости: ', horse['22.Об.белок_2'].mean())

Cреднее арифметическое, Ректальная температура:  38.166527196652744
Cреднее арифметическое, Пульс:  71.93454545454546
Cреднее арифметическое, Назогастральный рефлюкс, PH:  4.707547169811321
Cреднее арифметическое, Общий белок брюшной полости:  3.0196078431372553


Значение среднего арифметического ректальной температуры чуть выше нормы, пульса - превышают норму в два раза, показатели  назогастрального рефлюкса (PH) и общего белка брюшной полости находятся в пределах нормы.

В тоже время показатели среднего арифметического для анализа показателей здоровья по группе являются не информативными.

In [33]:
## Значение моды

print('Значение моды, операция: ', horse['1.Операция'].round().mode()[0])
print('Значение моды, возраст: ', horse['2.Возраст'].round().mode()[0])
print('Значение моды, Ректальная температура: ', horse['4.T_рек'].round().mode()[0])
print('Значение моды, Пульс: ', horse['5.Пульс'].round().mode()[0])
print('Значение моды, слизистые оболочки: ', horse['9.Слизистая'].round().mode()[0])
print('Значение моды, вздутие живота: ', horse['13.Вздутие'].round().mode()[0])
print('Значение моды, назогастральный рефлюкс, PH: ', horse['16.Рефлюкс_PH'].round().mode()[0])
print('Значение моды, общий белок брюшной полости: ', horse['22.Об.белок_2'].round().mode()[0])
print('Значение моды, результат: ', horse['23.Результат'].round().mode()[0])

Значение моды, операция:  1.0
Значение моды, возраст:  1
Значение моды, Ректальная температура:  38.0
Значение моды, Пульс:  48.0
Значение моды, слизистые оболочки:  1.0
Значение моды, вздутие живота:  1.0
Значение моды, назогастральный рефлюкс, PH:  6.0
Значение моды, общий белок брюшной полости:  2.0
Значение моды, результат:  1.0


Моды всех показателей укладываются в нормальные значения, при этом наибольшему количеству были проведены операции. Т.о.  показатель моды для анализа здоровья также является не информативным.

In [32]:
## медиана
print('Значение медианы, Ректальная температура: ', horse['4.T_рек'].median())
print('Значение медианы, Пульс: ', horse['5.Пульс'].median())
print('Значение медианы, слизистые оболочки: ', horse['9.Слизистая'].median())
print('Значение медианы, назогастральный рефлюкс, PH: ', horse['16.Рефлюкс_PH'].median())
print('Значение медианы, общий белок брюшной полости: ', horse['22.Об.белок_2'].median())

Значение медианы, Ректальная температура:  38.2
Значение медианы, Пульс:  64.0
Значение медианы, слизистые оболочки:  3.0
Значение медианы, назогастральный рефлюкс, PH:  5.0
Значение медианы, общий белок брюшной полости:  2.25


Медианы по выборке показывают незначительное повышение температуры, учащенное сердцебиение, бледность слизистых (что характерно для плохого кровотока), чуть выше нормы показатели  назогастрального рефлюкса и наличие белка в жидкости брюшной полости.

In [35]:
## СКО

print('СКО, Ректальная температура: ', horse['4.T_рек'].std())
print('СКО, Пульс: ', horse['5.Пульс'].std())
print('СКО, назогастральный рефлюкс, PH: ', horse['16.Рефлюкс_PH'].std())
print('СКО, общий белок брюшной полости: ', horse['22.Об.белок_2'].std())

СКО, Ректальная температура:  0.7335083691537396
СКО, Пульс:  28.68052200365475
СКО, назогастральный рефлюкс, PH:  1.9823111081419562
СКО, общий белок брюшной полости:  1.9685668112194978


In [40]:
## Дисперсия

print('Дисперсия, Ректальная температура: ', horse['4.T_рек'].var())
print('Дисперсия, Пульс: ', horse['5.Пульс'].var())
print('Дисперсия, назогастральный рефлюкс, PH: ', horse['16.Рефлюкс_PH'].var())
print('Дисперсия, общий белок брюшной полости: ', horse['22.Об.белок_2'].var())

Дисперсия, Ректальная температура:  0.5380345276185787
Дисперсия, Пульс:  822.5723424021243
Дисперсия, назогастральный рефлюкс, PH:  3.92955732946299
Дисперсия, общий белок брюшной полости:  3.8752552902349024


СКО и дисперсия также показывают незначительные отклонения в температуре и значительные отклонения в показателях пульса

In [46]:
## Квантили

print('Ректальная температура: I квантиль', horse['4.T_рек'].quantile(0.25), 'II квантиль', horse['4.T_рек'].quantile(), 'III квантиль', horse['4.T_рек'].quantile(0.75))
print('межквартильный размах: ', horse['4.T_рек'].quantile(0.75) - horse['4.T_рек'].quantile(0.25))

print('Пульс: I квантиль', horse['5.Пульс'].quantile(0.25), 'II квантиль', horse['5.Пульс'].quantile(), 'III квантиль', horse['5.Пульс'].quantile(0.75))
print('межквартильный размах: ', horse['16.Рефлюкс_PH'].quantile(0.75) - horse['16.Рефлюкс_PH'].quantile(0.25))

print('назогастральный рефлюкс, PH: I квантиль', horse['16.Рефлюкс_PH'].quantile(0.25), 'II квантиль', horse['16.Рефлюкс_PH'].quantile(), 'III квантиль', horse['16.Рефлюкс_PH'].quantile(0.75))
print('межквартильный размах: ', horse['16.Рефлюкс_PH'].quantile(0.75) - horse['16.Рефлюкс_PH'].quantile(0.25))

print('Общий белок брюшной полости: I квантиль', horse['22.Об.белок_2'].quantile(0.25), 'II квантиль', horse['22.Об.белок_2'].quantile(), 'III квантиль', horse['22.Об.белок_2'].quantile(0.75))
print('межквартильный размах: ', horse['22.Об.белок_2'].quantile(0.75) - horse['22.Об.белок_2'].quantile(0.25))

Ректальная температура: I квантиль 37.8 II квантиль 38.2 III квантиль 38.5
межквартильный размах:  0.7000000000000028
Пульс: I квантиль 48.0 II квантиль 64.0 III квантиль 88.0
межквартильный размах:  3.5
назогастральный рефлюкс, PH: I квантиль 3.0 II квантиль 5.0 III квантиль 6.5
межквартильный размах:  3.5
Общий белок брюшной полости: I квантиль 2.0 II квантиль 2.25 III квантиль 3.9
межквартильный размах:  1.9


In [47]:
## Describe
horse.describe()

1.Операция  2.Возраст    3.Больницa     4.T_рек     5.Пульс  \
count  298.000000  299.00000  2.990000e+02  239.000000  275.000000   
mean     1.395973    1.64214  1.087748e+06   38.166527   71.934545   
std      0.489881    2.17730  1.532026e+06    0.733508   28.680522   
min      1.000000    1.00000  5.184760e+05   35.400000   30.000000   
25%      1.000000    1.00000  5.289040e+05   37.800000   48.000000   
50%      1.000000    1.00000  5.303100e+05   38.200000   64.000000   
75%      2.000000    1.00000  5.347360e+05   38.500000   88.000000   
max      2.000000    9.00000  5.305629e+06   40.800000  184.000000   

        6.Дыхание     7.T_кон   8.Пульс_п  9.Слизистая  10.t _капилляры  ...  \
count  241.000000  243.000000  230.000000   253.000000       267.000000  ...   
mean    30.427386    2.345679    2.013043     2.853755         1.303371  ...   
std     17.678256    1.046369    1.042672     1.620294         0.476623  ...   
min      8.000000    1.000000    1.000000     1.000000         1.000000  ...   
25%     18.000000    1.000000    1.000000     1.000000         1.000000  ...   
50%     24.000000    3.000000    1.500000     3.000000         1.000000  ...   
75%     36.000000    3.000000    3.000000     4.000000         2.000000  ...   
max     96.000000    4.000000    4.000000     6.000000         3.000000  ...   

       19.Эритроциты  20.Об.белок_1  21.Бр.полость  22.Об.белок_2  \
count     270.000000     266.000000     135.000000     102.000000   
mean       46.300000      24.517293       2.037037       3.019608   
std        10.438384      27.509056       0.804905       1.968567   
min        23.000000       3.300000       1.000000       0.100000   
25%        38.000000       6.500000       1.000000       2.000000   
50%        45.000000       7.500000       2.000000       2.250000   
75%        52.000000      57.000000       3.000000       3.900000   
max        75.000000      89.000000       3.000000      10.100000   

       23.Результат  24.Х_поражение  25.Тип_пораж1  26.Тип_пораж2  \
count    298.000000      299.000000     299.000000     299.000000   
mean       1.550336        1.361204    3632.321070      90.528428   
std        0.737967        0.481155    5390.356722     650.637139   
min        1.000000        1.000000       0.000000       0.000000   
25%        1.000000        1.000000    2111.500000       0.000000   
50%        1.000000        1.000000    2322.000000       0.000000   
75%        2.000000        2.000000    3209.000000       0.000000   
max        3.000000        2.000000   41110.000000    7111.000000   

       27.Тип_пораж3  28.Паталогии  
count     299.000000    299.000000  
mean        7.387960      1.668896  
std       127.749768      0.471399  
min         0.000000      1.000000  
25%         0.000000      1.000000  
50%         0.000000      2.000000  
75%         0.000000      2.000000  
max      2209.000000      2.000000  

[8 rows x 28 columns]

### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [7]:
## поиск выброса через межквартильный размах для столбца 4:ректальная температура

q1 = horse['4.T_рек'].quantile(0.25)
q3 = horse['4.T_рек'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = horse[horse['4.T_рек'].between(lower_bound, upper_bound, inclusive=True)].sort_values('4.T_рек')
remove_outliers

1.Операция  2.Возраст  3.Больницa  4.T_рек  5.Пульс  6.Дыхание  7.T_кон  \
112         1.0          1      527933     36.8     60.0       28.0      NaN   
276         2.0          1      528620     36.9     50.0       40.0      2.0   
291         1.0          1      530034     37.0     66.0       20.0      1.0   
69          2.0          1      528214     37.0     60.0       20.0      3.0   
262         1.0          1      528743     37.1     75.0       36.0      NaN   
..          ...        ...         ...      ...      ...        ...      ...   
119         1.0          1      528682     39.4     54.0       66.0      1.0   
286         2.0          9     5305129     39.5     84.0       30.0      NaN   
161         2.0          1      535176     39.5     60.0       10.0      3.0   
230         2.0          1      530028     39.5     92.0       28.0      3.0   
116         2.0          1     5290482     39.5      NaN        NaN      3.0   

     8.Пульс_п  9.Слизистая  10.t _капилляры  ...  19.Эритроциты  \
112        NaN          NaN              NaN  ...            NaN   
276        3.0          3.0              1.0  ...           37.5   
291        3.0          2.0              1.0  ...           35.0   
69         NaN          NaN              1.0  ...           43.0   
262        NaN          3.0              2.0  ...           48.0   
..         ...          ...              ...  ...            ...   
119        1.0          2.0              1.0  ...           39.0   
286        NaN          NaN              1.0  ...           28.0   
161        NaN          NaN              2.0  ...           38.0   
230        3.0          6.0              1.0  ...           72.0   
116        3.0          4.0              2.0  ...            NaN   

     20.Об.белок_1  21.Бр.полость  22.Об.белок_2  23.Результат  \
112            NaN            NaN           10.0           2.0   
276            6.5            NaN            NaN           1.0   
291            6.9            2.0            NaN           2.0   
69             7.6            NaN            NaN           3.0   
262            7.4            3.0            3.2           2.0   
..             ...            ...            ...           ...   
119            6.0            2.0            NaN           1.0   
286            5.0            NaN            NaN           1.0   
161           56.0            1.0            NaN           1.0   
230            6.4            NaN            3.6           2.0   
116            6.7            1.0            NaN           3.0   

     24.Х_поражение  25.Тип_пораж1  26.Тип_пораж2  27.Тип_пораж3  28.Паталогии  
112               1           3205              0              0             1  
276               2           3111              0              0             2  
291               1          31110              0              0             2  
69                1           2209              0              0             1  
262               1          12208              0              0             1  
..              ...            ...            ...            ...           ...  
119               1           2206              0              0             1  
286               2              0              0              0             2  
161               2              0              0              0             2  
230               2           3300              0              0             2  
116               1           4205              0              0             2  

[225 rows x 28 columns]

In [8]:
# что это за выброс?
pd.concat([horse, remove_outliers]).drop_duplicates(keep=False)

1.Операция  2.Возраст  3.Больницa  4.T_рек  5.Пульс  6.Дыхание  7.T_кон  \
4           2.0          1      528355      NaN      NaN        NaN      2.0   
6           1.0          1      529607      NaN     60.0        NaN      3.0   
7           2.0          1      530051      NaN     80.0       36.0      3.0   
15          1.0          9     5301219      NaN    128.0       36.0      3.0   
19          1.0          1      530157     39.9     72.0       60.0      1.0   
..          ...        ...         ...      ...      ...        ...      ...   
281         1.0          1      527544      NaN     70.0       16.0      3.0   
287         1.0          1      529428      NaN      NaN        NaN      NaN   
292         1.0          1      534004      NaN     78.0       24.0      3.0   
294         1.0          1      533886      NaN    120.0       70.0      4.0   
297         1.0          1      530612     36.5    100.0       24.0      3.0   

     8.Пульс_п  9.Слизистая  10.t _капилляры  ...  19.Эритроциты  \
4          1.0          3.0              1.0  ...            NaN   
6          NaN          NaN              1.0  ...           44.0   
7          4.0          3.0              1.0  ...           38.0   
15         3.0          4.0              2.0  ...           53.0   
19         1.0          5.0              2.0  ...           46.0   
..         ...          ...              ...  ...            ...   
281        4.0          5.0              2.0  ...           60.0   
287        NaN          NaN              NaN  ...            NaN   
292        3.0          3.0              1.0  ...           43.0   
294        NaN          4.0              2.0  ...           55.0   
297        3.0          3.0              1.0  ...           50.0   

     20.Об.белок_1  21.Бр.полость  22.Об.белок_2  23.Результат  \
4              NaN            NaN            NaN           1.0   
6              8.3            NaN            NaN           2.0   
7              6.2            NaN            NaN           3.0   
15             7.8            3.0            4.7           2.0   
19             6.1            2.0            NaN           1.0   
..             ...            ...            ...           ...   
281            7.5            NaN            NaN           2.0   
287            NaN            NaN            NaN           1.0   
292           62.0            NaN            2.0           3.0   
294           65.0            NaN            NaN           3.0   
297            6.0            3.0            3.4           1.0   

     24.Х_поражение  25.Тип_пораж1  26.Тип_пораж2  27.Тип_пораж3  28.Паталогии  
4                 2              0              0              0             2  
6                 1           2208              0              0             2  
7                 1           3205              0              0             2  
15                2           1400              0              0             1  
19                1           2111              0              0             2  
..              ...            ...            ...            ...           ...  
281               1           1400              0              0             2  
287               1           2124              0              0             2  
292               2           2209              0              0             2  
294               2           3205              0              0             2  
297               1           2208              0              0             1  

[74 rows x 28 columns]

In [9]:
## поиск выброса через межквартильный размах для столбца 5:пульс

q1 = horse['5.Пульс'].quantile(0.25)
q3 = horse['5.Пульс'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = horse[horse['5.Пульс'].between(lower_bound, upper_bound, inclusive=True)].sort_values('5.Пульс')
remove_outliers

1.Операция  2.Возраст  3.Больницa  4.T_рек  5.Пульс  6.Дыхание  7.T_кон  \
129         1.0          1      528964     37.8     30.0       12.0      NaN   
231         1.0          1     5262543     38.5     30.0       18.0      NaN   
275         1.0          1      534787     38.0     36.0       16.0      3.0   
202         2.0          1      529685     37.2     36.0        9.0      1.0   
241         1.0          1      530354      NaN     36.0       32.0      1.0   
..          ...        ...         ...      ...      ...        ...      ...   
147         1.0          1      529272     38.3    132.0        NaN      NaN   
134         1.0          9     5287179     38.1    136.0       48.0      3.0   
43          1.0          1      535407     35.4    140.0       24.0      3.0   
102         1.0          9     5292489     38.0    140.0       68.0      1.0   
38          1.0          9     5277409     39.2    146.0       96.0      NaN   

     8.Пульс_п  9.Слизистая  10.t _капилляры  ...  19.Эритроциты  \
129        NaN          NaN              NaN  ...            NaN   
231        NaN          NaN              NaN  ...           40.0   
275        1.0          1.0              1.0  ...           37.0   
202        1.0          1.0              1.0  ...           35.0   
241        1.0          4.0              1.0  ...           41.0   
..         ...          ...              ...  ...            ...   
147        3.0          6.0              2.0  ...           57.0   
134        3.0          3.0              1.0  ...           33.0   
43         3.0          4.0              2.0  ...           57.0   
102        1.0          1.0              1.0  ...           39.0   
38         NaN          NaN              NaN  ...            NaN   

     20.Об.белок_1  21.Бр.полость  22.Об.белок_2  23.Результат  \
129            NaN            NaN            NaN           2.0   
231            7.7            NaN            NaN           1.0   
275           75.0            2.0            1.0           3.0   
202            5.7            NaN            NaN           1.0   
241            5.9            NaN            NaN           2.0   
..             ...            ...            ...           ...   
147            8.0            NaN            5.2           1.0   
134            4.9            2.0            2.9           2.0   
43            69.0            3.0            2.0           3.0   
102            5.3            NaN            NaN           1.0   
38             NaN            NaN            NaN           2.0   

     24.Х_поражение  25.Тип_пораж1  26.Тип_пораж2  27.Тип_пораж3  28.Паталогии  
129               1           2113              0              0             2  
231               1           2113              0              0             2  
275               2           3111              0              0             2  
202               2          31110              0              0             2  
241               1           2205              0              0             2  
..              ...            ...            ...            ...           ...  
147               1           2112              0              0             1  
134               1           2205              0              0             1  
43                1           3205              0              0             2  
102               1           7111              0              0             2  
38                1           2113              0              0             2  

[270 rows x 28 columns]

In [10]:
# что это за выброс?
pd.concat([horse, remove_outliers]).drop_duplicates(keep=False)

1.Операция  2.Возраст  3.Больницa  4.T_рек  5.Пульс  6.Дыхание  7.T_кон  \
2           1.0          9     5290409     39.1    164.0       84.0      4.0   
4           2.0          1      528355      NaN      NaN        NaN      2.0   
27          1.0          1     5279442      NaN      NaN        NaN      NaN   
40          2.0          9     5288249     39.0    150.0       72.0      NaN   
51          2.0          1      529483      NaN      NaN        NaN      1.0   
54          1.0          9     5282839     38.6    160.0       20.0      3.0   
55          1.0          1      528872      NaN      NaN        NaN      NaN   
57          1.0          1      528298      NaN      NaN       20.0      4.0   
73          1.0          9     5292929      NaN      NaN        NaN      NaN   
77          1.0          1      530693      NaN      NaN        NaN      3.0   
82          1.0          1     5279822     38.0      NaN       24.0      3.0   
92          2.0          1      530310      NaN      NaN        NaN      3.0   
114         2.0          1      533723      NaN      NaN       40.0      3.0   
116         2.0          1     5290482     39.5      NaN        NaN      3.0   
125         1.0          1      530384     38.7      NaN        NaN      3.0   
135         1.0          1      530693      NaN      NaN        NaN      3.0   
150         1.0          1      529399     39.3      NaN        NaN      4.0   
158         2.0          1      528134      NaN      NaN       12.0      1.0   
159         1.0          1      527916      NaN      NaN        NaN      NaN   
173         2.0          1      518476      NaN      NaN        NaN      3.0   
174         1.0          1      527929      NaN      NaN        NaN      NaN   
181         1.0          1     5279822     38.0      NaN       24.0      3.0   
203         1.0          1      529528     39.2      NaN       23.0      3.0   
206         1.0          1      530297      NaN      NaN        NaN      NaN   
215         2.0          1      528977      NaN      NaN        NaN      NaN   
226         1.0          9     5289419      NaN      NaN        NaN      NaN   
254         1.0          9     5294539     38.8    184.0       84.0      1.0   
274         1.0          9     5297159     38.8    150.0       50.0      1.0   
287         1.0          1      529428      NaN      NaN        NaN      NaN   

     8.Пульс_п  9.Слизистая  10.t _капилляры  ...  19.Эритроциты  \
2          1.0          6.0              2.0  ...           48.0   
4          1.0          3.0              1.0  ...            NaN   
27         NaN          NaN              NaN  ...            NaN   
40         NaN          NaN              NaN  ...           47.0   
51         1.0          3.0              1.0  ...           43.0   
54         NaN          5.0              1.0  ...           38.0   
55         NaN          NaN              NaN  ...           24.0   
57         3.0          3.0              NaN  ...           53.0   
73         NaN          NaN              NaN  ...           37.0   
77         3.0          3.0              3.0  ...           46.0   
82         3.0          6.0              2.0  ...           68.0   
92         3.0          1.0              1.0  ...           38.0   
114        1.0          1.0              1.0  ...           45.0   
116        3.0          4.0              2.0  ...            NaN   
125        1.0          3.0              1.0  ...           33.0   
135        3.0          3.0              2.0  ...           46.0   
150        3.0          6.0              2.0  ...           75.0   
158        1.0          2.0              1.0  ...           44.0   
159        NaN          NaN              NaN  ...           35.0   
173        4.0          6.0              NaN  ...            NaN   
174        NaN          NaN              NaN  ...            NaN   
181        3.0          6.0              2.0  ...           68.0   
203        1.0          3.0             

In [22]:
## поиск выброса через межквартильный размах для столбца 16:назогастральный рефлюкс, PH

q1 = horse['16.Рефлюкс_PH'].quantile(0.25)
q3 = horse['16.Рефлюкс_PH'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = horse[horse['16.Рефлюкс_PH'].between(lower_bound, upper_bound, inclusive=True)].sort_values('16.Рефлюкс_PH')
remove_outliers

1.Операция  2.Возраст  3.Больницa  4.T_рек  5.Пульс  6.Дыхание  7.T_кон  \
96          1.0          1      534145     38.1     88.0       32.0      3.0   
89          2.0          1      534197     38.0     52.0       16.0      NaN   
130         2.0          1      530670     37.6     88.0       36.0      3.0   
171         1.0          1      535292     38.5    129.0       48.0      3.0   
275         1.0          1      534787     38.0     36.0       16.0      3.0   
156         1.0          1      533928     37.8     60.0       24.0      1.0   
151         1.0          1      533836      NaN    100.0        NaN      3.0   
23          1.0          1      533692     38.1     60.0       12.0      3.0   
145         1.0          1      534115     38.2     48.0        NaN      1.0   
143         1.0          1      534519     37.1     84.0       40.0      3.0   
249         1.0          1      535043     39.3    100.0       51.0      4.0   
65          2.0          1      534784     37.9     72.0        NaN      1.0   
188         1.0          1     5299603     38.3     60.0       16.0      3.0   
211         1.0          9     5294369     38.9    120.0       30.0      1.0   
9           1.0          1      528548     38.1     66.0       12.0      3.0   
269         1.0          1      528570     37.5     60.0       50.0      3.0   
241         1.0          1      530354      NaN     36.0       32.0      1.0   
150         1.0          1      529399     39.3      NaN        NaN      4.0   
29          2.0          1      529475     37.7     96.0       30.0      3.0   
176         1.0          1      523190      NaN     68.0       14.0      NaN   
134         1.0          9     5287179     38.1    136.0       48.0      3.0   
221         1.0          1      528800      NaN    112.0       13.0      4.0   
14          1.0          1      530233     37.6     96.0       48.0      3.0   
69          2.0          1      528214     37.0     60.0       20.0      3.0   
262         1.0          1      528743     37.1     75.0       36.0      NaN   
154         2.0          1      528247     38.0     48.0       20.0      3.0   
2           1.0          9     5290409     39.1    164.0       84.0      4.0   
78          1.0          1      527463     37.5     96.0       18.0      1.0   
180         2.0          1      528305     37.8     48.0       14.0      NaN   
266         2.0          1      527829     38.3     54.0       18.0      3.0   
116         2.0          1     5290482     39.5      NaN        NaN      3.0   
20          2.0          1      529340     38.4     48.0       16.0      1.0   
67          2.0          1      529777      NaN     52.0       16.0      1.0   
83          1.0          1      529849     37.8     60.0       80.0      1.0   
103         1.0          1      527563     37.8     52.0       24.0      1.0   
30          2.0          1      530242     37.2    108.0       12.0      3.0   
147         1.0          1      529272     38.3    132.0        NaN      NaN   
268         1.0          1      527883     39.3     64.0       90.0      2.0   
245         1.0          1      530366     37.6     68.0       32.0      3.0   
146         2.0          1      530001     38.0     44.0       12.0      2.0   
167         1.0          1      530254     38.4    104.0       40.0      1.0   
179         1.0          1      530239     38.4     84.0       30.0      3.0   
234         1.0          1      528469     38.1     52.0       24.0      1.0   
90          2.0          1      529461     40.3    114.0       36.0      3.0   
72          1.0          1      528469     38.1     52.0       24.0      1.0   
63          1.0          1      529667     39.0     64.0       36.0      3.0   
53          2.0          1      529461     40.3    114.0       36.0      3.0   
47          1.0          1      528890     38.9     80.0       44.0      3.0   
276         2.0          1      528620     36.9     50.0       40.0      2.0   
290  

In [23]:
# что это за выброс?
pd.concat([horse['16.Рефлюкс_PH'], remove_outliers]).drop_duplicates(keep=False)


0  1.Операция  10.t _капилляры  11.Боль  12.Перистальтика  13.Вздутие  \
12   7.2         NaN              NaN      NaN               NaN         NaN   
30   6.0         NaN              NaN      NaN               NaN         NaN   
103  5.7         NaN              NaN      NaN               NaN         NaN   
130  1.5         NaN              NaN      NaN               NaN         NaN   
134  4.4         NaN              NaN      NaN               NaN         NaN   
..   ...         ...              ...      ...               ...         ...   
276  NaN         2.0              1.0      1.0               3.0         2.0   
290  NaN         1.0              1.0      2.0               3.0         3.0   
12   NaN         2.0              1.0      1.0               3.0         2.0   
278  NaN         2.0              1.0      2.0               4.0         3.0   
256  NaN         1.0              1.0      NaN               4.0         NaN   

     14.Зонд  15.Рефлюкс  16.Рефлюкс_PH  17.Кал  ...  26.Тип_пораж2  \
12       NaN         NaN            NaN     NaN  ...            NaN   
30       NaN         NaN            NaN     NaN  ...            NaN   
103      NaN         NaN            NaN     NaN  ...            NaN   
130      NaN         NaN            NaN     NaN  ...            NaN   
134      NaN         NaN            NaN     NaN  ...            NaN   
..       ...         ...            ...     ...  ...            ...   
276      3.0         1.0            7.0     NaN  ...            0.0   
290      2.0         2.0            7.0     3.0  ...            0.0   
12       3.0         NaN            7.2     1.0  ...            0.0   
278      2.0         2.0            7.5     NaN  ...            0.0   
256      NaN         3.0            7.5     NaN  ...            0.0   

     27.Тип_пораж3  28.Паталогии  3.Больницa  4.T_рек  5.Пульс  6.Дыхание  \
12             NaN           NaN         NaN      NaN      NaN        NaN   
30             NaN           NaN         NaN      NaN      NaN        NaN   
103            NaN           NaN         NaN      NaN      NaN        NaN   
130            NaN           NaN         NaN      NaN      NaN        NaN   
134            NaN           NaN         NaN      NaN      NaN        NaN   
..             ...           ...         ...      ...      ...        ...   
276            0.0           2.0    528620.0     36.9     50.0       40.0   
290            0.0           2.0    528890.0     38.9     80.0       44.0   
12             0.0           1.0   5291329.0     38.0     92.0       28.0   
278            0.0           1.0    529172.0     38.2     56.0       40.0   
256            0.0           1.0    529812.0     38.7     96.0       28.0   

     7.T_кон  8.Пульс_п  9.Слизистая  
12       NaN        NaN          NaN  
30       NaN        NaN          NaN  
103      NaN        NaN          NaN  
130      NaN        NaN          NaN  
134      NaN        NaN          NaN  
..       ...        ...          ...  
276      2.0        3.0          3.0  
290      3.0        3.0          3.0  
12       1.0        1.0          2.0  
278      4.0        3.0          1.0  
256      3.0        3.0          4.0  

[63 rows x 29 columns]

In [27]:
## поиск выброса через межквартильный размах для столбца 22:общий белок брюшной полости

q1 = horse['22.Об.белок_2'].quantile(0.25)
q3 = horse['22.Об.белок_2'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = horse[horse['22.Об.белок_2'].between(lower_bound, upper_bound, inclusive=True)].sort_values('22.Об.белок_2')
remove_outliers

1.Операция  2.Возраст  3.Больницa  4.T_рек  5.Пульс  6.Дыхание  7.T_кон  \
40          2.0          9     5288249     39.0    150.0       72.0      NaN   
298         1.0          1      534618     37.2     40.0       20.0      NaN   
121         2.0          1      534579     38.7     40.0       28.0      2.0   
126         1.0          1     5262541     37.4     50.0       32.0      3.0   
127         1.0          1      535330     37.4     84.0       20.0      NaN   
..          ...        ...         ...      ...      ...        ...      ...   
288         1.0          1      529126     38.0     50.0       36.0      NaN   
2           1.0          9     5290409     39.1    164.0       84.0      4.0   
170         2.0          1      527677     39.0     86.0       16.0      3.0   
61          1.0          1      528630     37.8     88.0       22.0      2.0   
32          1.0          1      529663     38.2     64.0       28.0      1.0   

     8.Пульс_п  9.Слизистая  10.t _капилляры  ...  19.Эритроциты  \
40         NaN          NaN              NaN  ...           47.0   
298        NaN          NaN              NaN  ...           36.0   
121        1.0          1.0              1.0  ...           39.0   
126        3.0          NaN              1.0  ...           45.0   
127        NaN          3.0              1.0  ...           31.0   
..         ...          ...              ...  ...            ...   
288        1.0          1.0              1.0  ...           39.0   
2          1.0          6.0              2.0  ...           48.0   
170        3.0          5.0              NaN  ...           68.0   
61         1.0          2.0              1.0  ...           64.0   
32         1.0          1.0              1.0  ...           49.0   

     20.Об.белок_1  21.Бр.полость  22.Об.белок_2  23.Результат  \
40             8.5            NaN            0.1           1.0   
298           62.0            1.0            1.0           3.0   
121           62.0            1.0            1.0           1.0   
126            7.9            2.0            1.0           1.0   
127           61.0            NaN            1.0           3.0   
..             ...            ...            ...           ...   
288            6.6            1.0            5.3           1.0   
2              7.2            3.0            5.3           2.0   
170            5.8            3.0            6.0           2.0   
61             8.0            1.0            6.0           2.0   
32             8.6            2.0            6.6           1.0   

     24.Х_поражение  25.Тип_пораж1  26.Тип_пораж2  27.Тип_пораж3  28.Паталогии  
40                1           9400              0              0             1  
298               2           6112              0              0             2  
121               2              0              0              0             2  
126               1           2208              0              0             1  
127               2           2208              0              0             2  
..              ...            ...            ...            ...           ...  
288               1           8400              0              0             1  
2                 1           2208              0              0             1  
170               1           5400              0              0             1  
61                1           3205              0              0             1  
32                1           2208              0              0             1  

[95 rows x 28 columns]

In [30]:
# что это за выброс?
pd.concat([horse, remove_outliers]).drop_duplicates(keep=False)

1.Операция  2.Возраст  3.Больницa  4.T_рек  5.Пульс  6.Дыхание  7.T_кон  \
1           2.0          1      530334     38.3     40.0       24.0      1.0   
3           2.0          1      530255     37.3    104.0       35.0      NaN   
4           2.0          1      528355      NaN      NaN        NaN      2.0   
5           1.0          1      526802     37.9     48.0       16.0      1.0   
6           1.0          1      529607      NaN     60.0        NaN      3.0   
..          ...        ...         ...      ...      ...        ...      ...   
290         1.0          1      528890     38.9     80.0       44.0      3.0   
291         1.0          1      530034     37.0     66.0       20.0      1.0   
293         2.0          1      533902     38.5     40.0       16.0      1.0   
294         1.0          1      533886      NaN    120.0       70.0      4.0   
296         1.0          1      529386     37.5     72.0       30.0      4.0   

     8.Пульс_п  9.Слизистая  10.t _капилляры  ...  19.Эритроциты  \
1          1.0          3.0              1.0  ...           33.0   
3          NaN          6.0              2.0  ...           74.0   
4          1.0          3.0              1.0  ...            NaN   
5          1.0          1.0              1.0  ...           37.0   
6          NaN          NaN              1.0  ...           44.0   
..         ...          ...              ...  ...            ...   
290        3.0          3.0              1.0  ...           54.0   
291        3.0          2.0              1.0  ...           35.0   
293        1.0          1.0              1.0  ...           37.0   
294        NaN          4.0              2.0  ...           55.0   
296        3.0          4.0              1.0  ...           60.0   

     20.Об.белок_1  21.Бр.полость  22.Об.белок_2  23.Результат  \
1              6.7            NaN            NaN           1.0   
3              7.4            NaN            NaN           2.0   
4              NaN            NaN            NaN           1.0   
5              7.0            NaN            NaN           1.0   
6              8.3            NaN            NaN           2.0   
..             ...            ...            ...           ...   
290            6.5            3.0            NaN           2.0   
291            6.9            2.0            NaN           2.0   
293           67.0            NaN            NaN           1.0   
294           65.0            NaN            NaN           3.0   
296            6.8            NaN            NaN           2.0   

     24.Х_поражение  25.Тип_пораж1  26.Тип_пораж2  27.Тип_пораж3  28.Паталогии  
1                 2              0              0              0             1  
3                 2           4300              0              0             2  
4                 2              0              0              0             2  
5                 1           3124              0              0             2  
6                 1           2208              0              0             2  
..              ...            ...            ...            ...           ...  
290               1           7111              0              0             2  
291               1          31110              0              0             2  
293               2              0              0              0             2  
294               2           3205              0              0             2  
296               1           3205              0              0             2  

[204 rows x 28 columns]

Большое количество выбросов по показателям температуры и пульса связано с большим размахом в их значениях. Выбросы по показателям лабораторных данных (назогастрального рефлюкса и наличие белка в жидкости брюшной полости) в основном связано с отсутствием данных по этим показателям.

### Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [171]:
horse1 = horse[['1.Операция', '2.Возраст', '4.T_рек', '5.Пульс', '9.Слизистая', '13.Вздутие','16.Рефлюкс_PH', '22.Об.белок_2', '23.Результат']]
horse1.head()

1.Операция  2.Возраст  4.T_рек  5.Пульс  9.Слизистая  13.Вздутие  \
0         1.0          1     39.2     88.0          4.0         2.0   
1         2.0          1     38.3     40.0          3.0         1.0   
2         1.0          9     39.1    164.0          6.0         4.0   
3         2.0          1     37.3    104.0          6.0         NaN   
4         2.0          1      NaN      NaN          3.0         2.0   

   16.Рефлюкс_PH  22.Об.белок_2  23.Результат  
0            NaN            2.0           3.0  
1            NaN            NaN           1.0  
2            5.0            5.3           2.0  
3            NaN            NaN           2.0  
4            NaN            NaN           1.0

In [172]:
horse1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   1.Операция     298 non-null    float64
 1   2.Возраст      299 non-null    int64  
 2   4.T_рек        239 non-null    float64
 3   5.Пульс        275 non-null    float64
 4   9.Слизистая    253 non-null    float64
 5   13.Вздутие     243 non-null    float64
 6   16.Рефлюкс_PH  53 non-null     float64
 7   22.Об.белок_2  102 non-null    float64
 8   23.Результат   298 non-null    float64
dtypes: float64(8), int64(1)
memory usage: 21.1 KB


In [173]:
# Удалим только те строки, в которых как минимум 6/9 значений заполнено (при удалении 7/9 и 8/9 теряется много значений)
horse1.dropna(thresh=6).info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 259 entries, 0 to 298
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   1.Операция     258 non-null    float64
 1   2.Возраст      259 non-null    int64  
 2   4.T_рек        224 non-null    float64
 3   5.Пульс        253 non-null    float64
 4   9.Слизистая    241 non-null    float64
 5   13.Вздутие     231 non-null    float64
 6   16.Рефлюкс_PH  53 non-null     float64
 7   22.Об.белок_2  100 non-null    float64
 8   23.Результат   258 non-null    float64
dtypes: float64(8), int64(1)
memory usage: 20.2 KB


In [174]:
(horse1.isna().mean() * 100).round(2)

1.Операция        0.33
2.Возраст         0.00
4.T_рек          20.07
5.Пульс           8.03
9.Слизистая      15.38
13.Вздутие       18.73
16.Рефлюкс_PH    82.27
22.Об.белок_2    65.89
23.Результат      0.33
dtype: float64

Видно, что в столбце 16.Рефлюкс_PH  данные показателя назогастрального рефлюкса (PH) заполнены только на 18%,
заполнение данных средними значениями, либо модой в данном случае не логично, т.к. РН- метрия рефлюкса зависит от характера и тяжести заболевания. Можно предположить, что из-за сложности забора образцов у животных данные анализы практически не проводились. Для дальнейшего анализа из-за существенного отсутствия данных и невозможности их заполнения средними либо модой целесообразно удалить данный столбец.  


In [175]:
horse2 = horse1.drop(columns=['16.Рефлюкс_PH'], axis=1)

Данные в столбце "22:Общий белок брюшной полости" заполнены на 35%. В то же время анализ перитонеальной жидкости позволяет диагностировать патологическое состояние, которое является причиной скопления жидкости в брюшной полости (асцита) и/или воспаления брюшины (перитонита). Попробуем заполнить пропущенные данные, использую группировки по данным проведенной операции (вероятней всего именно перед операцией проводили лабораторные исследования), а также более распространенным показателям здоровья лошадей: ректальной температуры, пульса, вздутия и слизистых оболочек.
         

In [191]:
print(horse2.groupby(["1.Операция","4.T_рек", "5.Пульс","9.Слизистая", "13.Вздутие"])["22.Об.белок_2"].median())
horse3 = horse2
horse3["22.Об.белок_2"].fillna(horse2.groupby(["1.Операция", "4.T_рек", "5.Пульс","9.Слизистая", "13.Вздутие"], observed=True)["22.Об.белок_2"].transform('median'), inplace=True)


1.Операция  4.T_рек  5.Пульс  9.Слизистая  13.Вздутие
1.0         36.4     98.0     4.0          2.0           3.6
            36.5     78.0     1.0          1.0           1.0
                     100.0    3.0          3.0           3.4
            37.0     66.0     2.0          3.0           NaN
            37.1     40.0     4.0          1.0           NaN
                                                        ... 
2.0         39.4     110.0    6.0          3.0           NaN
            39.5     92.0     6.0          1.0           3.6
            39.6     108.0    6.0          3.0           2.6
            40.0     78.0     5.0          1.0           NaN
            40.3     114.0    1.0          3.0           4.5
Name: 22.Об.белок_2, Length: 173, dtype: float64


Аналогично заменим пропуски по показателям ректальной температуры, пульса, вздутия и слизистых оболочек, испоьлзую группировки по показателям здоровья и проведенным операциям

In [192]:
horse3['4.T_рек'].fillna(horse2.groupby(['1.Операция', '5.Пульс','9.Слизистая', '13.Вздутие'], observed=True)['4.T_рек'].transform('median'), inplace=True)

In [193]:
horse3['5.Пульс'].fillna(horse2.groupby(['1.Операция','4.T_рек','9.Слизистая', '13.Вздутие'], observed=True)['5.Пульс'].transform('median'), inplace=True)

In [194]:
horse3['9.Слизистая'].fillna(horse2.groupby(['1.Операция','4.T_рек', '5.Пульс', '13.Вздутие'], observed=True)['9.Слизистая'].transform('median'), inplace=True)

In [195]:
horse3['13.Вздутие'].fillna(horse2.groupby(['1.Операция','4.T_рек', '5.Пульс','9.Слизистая'], observed=True)['13.Вздутие'].transform('median'), inplace=True)

Данные по показателям проведенных операций и результата можно заполнить значением моды, т.к. эти показатели не слишком вариативны и данные по ним практически заполнены.

In [196]:
horse3['1.Операция'].fillna(horse2['1.Операция'].mode()[0], inplace=True)
horse3['1.Операция'].value_counts()

1.0    181
2.0    118
Name: 1.Операция, dtype: int64

In [197]:
horse3['23.Результат'].fillna(horse2['23.Результат'].mode()[0], inplace=True)
horse3['23.Результат'].value_counts()

1.0    179
2.0     76
3.0     44
Name: 23.Результат, dtype: int64